### ISSUE TITLE
Support for Arrow

### ISSUE THREAD

**Issue by rafapereirabr** [ipeaGIT/r5r#369]
Opening this issue with the suggestion that we include support for Arrow in r5r. 

[As documented on their website](https://arrow.apache.org/docs/index.html), Arrow specifies a standardized language-independent columnar memory format for flat and hierarchical data. This would mean two most obvious advantages: (1) passing data from Java to R (from R5 to r5r) would become seamless, (2) saving outputs in `.parquet` format. Both of these advantages would probably make r5r substantially faster, with more efficiency gains for large scale analyses.

There are robust implementations of Arrow in [Java](https://arrow.apache.org/cookbook/java/), [R](https://arrow.apache.org/cookbook/r/) and also in [Python](https://arrow.apache.org/cookbook/py/) (in case we want to implement this in r5py).

I'm not sure this could be done entirely within the Java side of `r5r` or whether it would require some change to R5 upstream. In any case, this might be something that the @conveyal would be interested in, since this would speed improve up the process of passing R5 results to interactive visualization in Conveyal Analysis.

**Comment by botanize** [comment]
I've been working with the csv output of `travel_time_matrix` for a large region and many of the csv files contain only a few lines, performance reading 30k files is poor. It would be a bit more work, but writing the matrices to parquet files, aggregating multiple `from_id`s would be hugely helpful.

**Comment by rafapereirabr** [comment]
yes, this could be a great improvment to te package. @botanize , if you're familiar with Java and would like to have a look at this, we would appreciate PR from collaborators

**Comment by rafapereirabr** [comment]
this would require major change upstream in R5. Closing the issue here

**Comment by BardyBard** [comment]
@rafapereirabr I have successfully transferred a travel times matrix from the little jar to R using arrow.

PR soon...

**Comment by rafapereirabr** [comment]
yahoo ! amazing. now the key question, how did it affect performance (computation time and memory) ?

**Comment by BardyBard** [comment]
@rafapereirabr I tried a square matrix, 1500 origins to 1500 destinations on my laptop.

10-15% improvement in total CPU time, and 20% improvement in real time passed. I assume the discrepancy between the real life time and CPU time is because arrow does a bit more multi threaded work compared to Marcus's code.

This is a good result I guess considering we're just changing the language translation code. But a bit disappointing nonetheless

**Comment by rafapereirabr** [comment]
Ok, so using Arrow to pass data from Java to R does not bring the big improvment I was expecting, but it's still something. I guess @mvpsaraiva code inside `java_to_dt()` was already pretty efficient then! @BardyBard , have you had the chance of testing it in larger travel time matrices?

**Comment by higgicd** [comment]
@BardyBard yes one hope was to enable analyses that might otherwise be too memory intensive, e.g. people running analyses in a class or workshop or something who have different memory specs. My tests with the current csv output suggest it doesn't really free up a ton of memory though, bulk is calculation as you note. 

But the other hope is just disk storage and calculation efficiency. E.g. I already make use of {purrr}-based solutions for iterating ttms for multiple modes or for segmenting origins and writing them out to parquet datasets that can be used later. Saves memory if you creatively batch origins and a ton of re-calculation time if you just load the finished ttms from disk later.
